File: backend/src/main/java/teamx/app/backend/models/Transaction.java
package teamx.app.backend.models;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import teamx.app.backend.utils.DTO.TransactionDTO;
import teamx.app.backend.utils.Model;

import java.sql.Date;

/**
 * Transaction entity
 * Represents a transaction
 *
 * @author Junior Javier Brito Perez
 * @see Product
 * @see Warehouse
 */
@Data
@Builder
@Entity(name = "Transactions")
@NoArgsConstructor
@AllArgsConstructor
public class Transaction implements Model<TransactionDTO> {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private int quantity;

    @Enumerated(EnumType.STRING)
    private Type transactionType;
    @ManyToOne
    @JsonIgnore
    private Product product;
    @JsonIgnore
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy")
    private Date transactionDate;
    @ManyToOne
    @JsonIgnore
    private Warehouse warehouse;
    @ManyToOne
    @JsonIgnore
    private Project project;
    @ManyToOne
    @JsonIgnore
    private Warehouse transferFrom;
    @ManyToOne
    @JsonIgnore
    private InventoryOrder inventoryOrder;


    public boolean isPositiveTransaction() {
        return this.getTransactionType() == Type.INVENTORY_ORDER ||
                (this.getTransactionType() == Type.ADJUSTMENT ||
                        this.getTransactionType() == Type.RETURN ||
                        this.getTransactionType() == Type.OTHER)
                        && this.getQuantity() > 0;
    }

    public enum Type {
        INVENTORY_ORDER, PROJECT_MATERIAL, TRANSFER, ADJUSTMENT,
        RETURN, DAMAGED, LOST, EXTRA_MATERIAL_FOR_PROJECT, OTHER
    }

    @Override
    public TransactionDTO toDTO() {
        return TransactionDTO.builder()
                .id(id)
                .quantity(quantity)
                .transactionType(transactionType.toString())
                .transactionDate(transactionDate)
                .productId(product != null ? product.getId() : null)
                .warehouseId(warehouse != null ? warehouse.getId() : null)
                .transferFromWarehouseId(transferFrom != null ? transferFrom.getId() : null)
                .projectId(project != null ? project.getId() : null)
                .orderId(inventoryOrder != null ? inventoryOrder.getId() : null)
                .build();
    }
}

File: frontend/src/models/Transaction.js
import {Model} from "@/models/Model";

export class Transaction extends Model{
    constructor(id, quantity, transactionType, transactionDate, productId, warehouseId, transferFromWarehouseId,
                projectId) {
        super();
        this.id = id;
        this.quantity = quantity;
        this.transactionType = transactionType;
        this.transactionDate = transactionDate;
        this.productId = productId;
        this.warehouseId = warehouseId;
        this.transferFromWarehouseId = transferFromWarehouseId;
        this.projectId = projectId;
    }

    static get CATEGORY() {
        return {
            INVENTORY_ORDER: "Order",
            PROJECT_MATERIAL: "Project Material",
            TRANSFER: 'Transfer',
            ADJUSTMENT: 'Adjustment',
            RETURN: 'Return',
            DAMAGED: 'Damaged',
            LOST: 'Lost',
            EXTRA_MATERIAL_FOR_PROJECT: 'Extra Project Material',
            OTHER: 'Other'
        }
    }

    static get getCategories() {
        let categories = [];
        Object.keys(this.CATEGORY).forEach(key => {
            categories.push({title: this.CATEGORY[key], value: key});
        });
        return categories;
    }

    static getTransactionFlow(transaction, warehouseId) {
        switch (transaction.transactionType) {
            case Transaction.CATEGORY.INVENTORY_ORDER:
            case Transaction.CATEGORY.RETURN:
                return 'IN';
            case Transaction.CATEGORY.PROJECT_MATERIAL:
            case Transaction.CATEGORY.EXTRA_MATERIAL_FOR_PROJECT:
            case Transaction.CATEGORY.DAMAGED:
            case Transaction.CATEGORY.LOST:
                return 'OUT';
            case Transaction.CATEGORY.TRANSFER:
                return transaction.warehouseId === warehouseId ? 'IN' : 'OUT';
            case Transaction.CATEGORY.ADJUSTMENT:
            case Transaction.CATEGORY.OTHER:
                return transaction.quantity > 0 ? 'IN' : 'OUT';
            default:
                throw new Error('Invalid transaction type for transaction: ' + transaction.id +
                    ' transaction type: ' + transaction.transactionType);
        }
    }

    static getTransactionFlowColor(transaction, warehouseId) {
        switch (this.getTransactionFlow(transaction, warehouseId)) {
            case 'IN':
                return 'green';
            case 'OUT':
                return 'red';
            default:
                console.error(transaction);
                console.error(warehouseId);
                console.error(Transaction.CATEGORY.INVENTORY_ORDER)
                throw new Error('Invalid transaction flow for transaction: ' + transaction.id +
                    ' transaction type: ' + transaction.transactionType);
        }
    }
}